### PR TITLE
Developer focus/repo spread

### PR DESCRIPTION
**Pull request by zlusch95**
Update of the repository spread calculation. 

@DieserFelix please check, if we need a weighted average again, as we calculate an average of averages, too. I am not quite sure about it.

Consider the following day spread data:
```
'2021-09-17': { gr2m: 9, 'web-flow': 3 },
'2021-09-22': { gr2m: 2, 'web-flow': 2 },
'2021-08-03': { gr2m: 2 },
'2021-08-09': { gr2m: 2 },
'2021-08-10': { gr2m: 1 },
```

This would lead to this average spread at those days

```
'2021-09-17': 6,
'2021-09-22': 2,
'2021-08-03': 2,
'2021-08-09': 2,
'2021-08-10': 1,
```

If we would weight the average now, we calculate `6*2/7 + 2*2/7 ...  = 3` and if we would not weight the average, we get `(6 + 3*2 + 1)/5 = 2.2`.

At the moment, I am doing the last one.
So our expression here is now an average spread per time items in a category instead of an average spread per dev (weighted) as in the total repo spread.

### CODE REVIEW

File: src/database/statistics/spreadUtil.ts [hunk 1]
```diff
@@ -71,110 +94,76 @@ export function getSpreadDates(
...
-      const devSpreadObj1 = timeSpreadPairs[currentDate];
-      const devSpreadObj2 = timeSpreadPairs[nextDate];
-      const mergedObj = { ...devSpreadObj1, ...devSpreadObj2 };
-      newTimeSpreadPairs[currentDate] = mergedObj;
-      // delete the unnessacary date and skip the next date
-      delete newTimeSpreadPairs[nextDate];
-      i += 2;
-    } else {
-      i += 1;
+function isRepoIDinDevSpreadArr(
```
Comment (DieserFelix) [inline src/database/statistics/spreadUtil.ts:155]: This name is confusing. I expect a function `isSomething` to check if its parameter fulfills some condition. I wouldn't expect it to have side effects like initializing an object.

**Comment by DieserFelix** [comment]
I don't think you should be calculating the repo spread on averages at all. To me, repo spread should be calculated analogous to an individual dev's spread:  
Let's say, three developers A, B and C contributed to at least repository R in week W1:  
| A | B | C |
| - | - | - |
| R | R | R |
| S | S |  |
| V |  |  |
| U | | |
| **4** | **2** | **1**|

And in week W2:  
| A | B | C |
| - | - | - |
| R | R | R |
| S | S |  |
| U | | |
| **3** | **2** | **1**|

The weekly spread of repo R for any week is the number of unique repositories contributed to by devs who also contributed to R. So:  
`weekSpread(R, W1) = 4`
`weekSpread(R, W2) = 3`

And the combined one is the average of W1 and W2:
`weekSpread(R) = 3.5`

My understanding might be wrong, though. Feel free to discuss.

**Comment by DieserFelix** [comment]
I forgot to mention: we should ignore bot users like web-flow or actions-user. They will naturally have contributed to many repositories but don't matter for us.

I don't know, however, if we can infer this from Github's API. If not, we should provide a list of bot names as a parameter.

**Comment by zlusch95** [comment]
> I don't think you should be calculating the repo spread on averages at all. To me, repo spread should be calculated analogous to an individual dev's spread 

> The weekly spread of repo R for any week is the number of unique repositories contributed to by devs who also contributed to R

I like your suggestion. I think, both models have their reasons. In your approach, we look at a repository in a more objective way, as for a "worst case example" like `spread(dev A) = 9` and `spread(dev B) = 1` would lead to `weekSpread(R) = 9`.
The second approach would lead to a `weekSpread(R) = 5` which indicates a high spread value as well. 
In analogy to individuel dev spread, I think your approach would be better. There, the calculation is in the same manner as we take an average in both calculations one time, but for the repo spread we don't need the last *weighted average* as we want du monitor only one repository. 

So, I think I would apply your suggestion then.

**Comment by zlusch95** [comment]
> I forgot to mention: we should ignore bot users like web-flow or actions-user. They will naturally have contributed to many repositories but don't matter for us.

> I don't know, however, if we can infer this from Github's API. If not, we should provide a list of bot names as a parameter.

Yes, exactly. We should exclude them. I don't know if there is a whole bunch of such bot users which all have different names or are they all the same and they provided by GitHub?

Anyway, I will have a look at the Api later. I will leave this to an issue, as we need it for the whole developer focus.

File: src/database/statistics/spreadUtil.ts (hunk 1 above)
Comment (zlusch95) [reply src/database/statistics/spreadUtil.ts:155]: Yeah, right. I changed it to `createSetAndAddRepoIds()` and I am referring with that to the calculation which you suggested down below.

**Comment by DieserFelix** [comment]
Looks good to me

**Pull request merged**